source/elements/oneDAL/source/data_management/data_sources.rst

```diff
@@ -1,7 +1,259 @@
+-------------------------
+`The following code sample <read_operation_template_>`_ shows the declaration for a
+read operation.
+
+.. code-block:: cpp
+   :name: read_operation_template
+
+   namespace oneapi::dal {
+
+   template <typename DatasetType, typename DataSource>
+   using read_args_t = /* implementation defined */;
+
+   template <typename DatasetType, typename DataSource>
+   using read_result_t = DatasetType;
+
+   template <typename DatasetType, typename DataSource>
+   read_result_t<DatasetType, DataSource> read(
+      sycl::queue& queue,
+      const DataSource& data_source,
+      const read_args_t<DatasetType, DataSource>& args);
+
+   } // namespace oneapi::dal
+
+
+Each operation shall satisfy the following requirements:
+
+- An operation shall accept three parameters in the following order:
+
+  + The SYCL* queue object.
+  + The data source.
+  + The `args object <args_>`_.
+
+- An operation shall return the `result object <result_>`_.
+
+- The ``read_args_t`` and ``read_result_t`` alias
+  templates shall be used for inference of the args and return types.
+
+------------------------
+Read operation shortcuts
+------------------------
+In order to make the code on user side less verbose, oneDAL defines the
+following overloaded functions called *shortcuts* for a read operation in addition
+to the general one described in section `Read operation definition
+<read_operation_definition_>`_.
+
+- Shortcut for execution on :term:`host <Host/Device>`. Performs the same
+  operation as the general function on host, but does not require passing the
+  queue explicitly.
+
+  .. code-block:: cpp
+
+     template <typename DatasetType, typename DataSource>
+     read_result_t<DatasetType, DataSource> read(
+        const DataSource& data_source,
+        const read_args_t<DatasetType, DataSource>& args);
+
+- Shortcut that allows omitting explicit args creation.
+
+  .. code-block:: cpp
+
+     template <typename DatasetType, typename DataSource, typename... Args>
+     read_result_t<DatasetType, DataSource> read(
+        sycl::queue& queue,
+        const DataSource& data_source,
+        Args&&... args);
+
+- Shortcut that allows omitting explicit queue and args creation. This is
+  combination of two previous shortcuts.
+
+  .. code-block:: cpp
+
+     template <typename DatasetType, typename DataSource, typename... Args>
+     read_result_t<DatasetType, DataSource> read(
+        const DataSource& data_source,
+        Args&&... args);
+
+
+.. _args:
+
+----
+Args
+----
+
+.. code-block:: cpp
+
+   namespace oneapi::dal::%DATA_SOURCE% {
+
+   template <typename DatasetType, typename DataSource>
+   class read_args {
+   public:
+      read_args(
+         const %PROPERTY_TYPE_1%& property_name_1,
+         const %PROPERTY_TYPE_2%& property_name_2,
+         /* more properties */
+      )
+      /* Getter & Setter for the property called `%PROPERTY_NAME_1%` */
+      descriptor& set_%PROPERTY_NAME_1%(%PROPERTY_TYPE_1% value);
```

Comment:
you use notation %DATA_SOURCE% and others but do not introduce it in the section. Namespaces/headers section has the example how to do that

Reply:
Fixed.